### PR TITLE
Add support for E PluralOperand

### LINKED ISSUE #228: Add support for `e` oprand to PluralOperands
In https://unicode-org.atlassian.net/browse/CLDR-12010 CLDR adds `e` operand. We should implement it as well.

### PR DESCRIPTION
**Pull request by zbraniecki**
Fixes #228

### CODE REVIEW

**Comment by zbraniecki** [comment]
This seems to be ready for review.

The issue is performance. It impacts it *a lot*. Like, `FromStr` benchmarks get 70-100% slower, and the `plurals/operands/overview` benchmark which benchmarks all types of inputs is getting ~18% hit (due to the FromStr being slower).

I could microoptimize it by refactoring to do a single iteration over the string, but since we want to switch `FromStr` to use `FixedDecimal` then maybe there's not that much value in it.

I'll consider trying just for fun, but since any potential optimizations are local to the `FromStr` impl, I think this PR is ready for review.

**Review by echeran — DISMISSED** [review]
LGTM. Seems pretty straightforward, and as you said, it makes sense to not think about the performance impact yet since you'll be moving stuff around including some of this into FixedDecimal as a followup anyways.

File: components/plurals/tests/fixtures/operands.json [hunk 1]
```diff
@@ -2,119 +2,127 @@
...
       "input": "2.0",
-      "output": [2, 2, 1, 0, 0, 0]
+      "output": [2, 2, 1, 0, 0, 0, 0]
+    },
+    {
+      "input": "2e3",
+      "output": [2, 2, 0, 0, 0, 0, 3]
+    },
+    {
+      "input": "2.0e12",
```
Comment (sffc) [inline components/plurals/tests/fixtures/operands.json:64]: Worth adding a test case for negative exponents? I think they're unlikely, but we should know how we behave (return an error? set to zero?)

**Comment by sffc** [comment]
Here's the official language on `e` vs `c`:

> Added new operand 'c' (with a synonym 'e') for languages like French (CLDR-12010)

So, in other words, we should name the field "c", but we should accept an alias "e" when parsing the rules.

https://sites.google.com/site/cldr/index/downloads/cldr-38?pli=1#TOC-Migration

File: components/plurals/src/operands.rs [hunk 2]
```diff
@@ -67,6 +70,8 @@ pub struct PluralOperands {
     pub f: u64,
     /// Visible fraction digits without trailing zeros
     pub t: u64,
+    /// Exponent
```
Comment (filmil) [inline components/plurals/src/operands.rs:73]: nit: consider being a bit more verbose about what the exponent means?

**Review by filmil — DISMISSED** [review]

File: components/plurals/tests/fixtures/operands.json (hunk 1 above)
Comment (zbraniecki) [reply components/plurals/tests/fixtures/operands.json:64]: I considered where to put it, but since UTS doesn't allow for them, I would only do this by extending our data driven `FromStr` tests, and we're planning to move that to `FixedDecimal` so I decided to not spend time on that. Instead, I hope we'll have such tests in `FromStr for FixedDecimal`.

Let me know if you believe we should invest time in rebuilding our `FromStr` test corpus or operands in Plurals.

File: components/plurals/src/rules/ast.rs [hunk 3]
```diff
@@ -264,6 +264,11 @@ pub enum Operand {
     F,
     /// Visible fraction digits without trailing zeros
     T,
+    /// Compact decimal exponent value:
+    ///   exponent of the power of 10 used in compact decimal formatting
+    C,
+    /// Currently, synonym for ‘c’. however, may be redefined in the future
+    E,
```
Comment (sffc) [inline components/plurals/src/rules/ast.rs:271]: I think I would only have C here.  I think `e` should only be used when parsing.

File: components/plurals/src/rules/lexer.rs [hunk 4]
```diff
@@ -165,8 +166,7 @@ impl<'l> Lexer<'l> {
                         Token::Ellipsis
                     }
                     b'~' => Token::Tilde,
-                    // `e` may come up also in samples for scientific notation.
-                    b'e' => Token::Operand(ast::Operand::E),
+                    b'e' => Token::E,
```
Comment (sffc) [inline components/plurals/src/rules/lexer.rs:169]:
```suggestion
                    b'e' => Token::Operand(ast::Operand::C),
```

File: components/plurals/tests/fixtures/rules.json [hunk 5]
```diff
@@ -121,10 +121,5 @@
     "rule": "e = 0 and i != 0 and i % 1000000 = 0 and v = 0 or e != 0..5 @integer 1000000, 1e6, 2e6, 3e6, 4e6, 5e6, 6e6, … @decimal 1.0000001e6, 1.1e6, 2.0000001e6, 2.1e6, 3.0000001e6, 3.1e6, …",
     "input": 1,
     "output": false
-  },
-  {
-    "rule": "c = 0 and i != 0 and i % 1000000 = 0 and v = 0 or c != 0..5 @integer 1000000, 1e6, 2e6, 3e6, 4e6, 5e6, 6e6, … @decimal 1.0000001e6, 1.1e6, 2.0000001e6, 2.1e6, 3.0000001e6, 3.1e6, …",
```
Comment (sffc) [inline components/plurals/tests/fixtures/rules.json]: Why did you delete this?

Comment (zbraniecki) [reply components/plurals/tests/fixtures/rules.json]: because we don't support `e = 0` anymore. Or, if we do, there's no way to do a rountrip if I can't store C vs. E as two AST nodes, which is what you suggested me to remove.

File: components/plurals/src/rules/lexer.rs (hunk 4 above)
Comment (zbraniecki) [reply components/plurals/src/rules/lexer.rs:169]: That will not work because lexer parses also samples and there you have `1e6` which has to come out as a `[Token::Number, Token::E, Token::Number]`.

**Review by sffc — APPROVED** [review]
